### PR TITLE
Use timeouts for FireEvent to avoid blocking bugs

### PR DESCRIPTION
**Pull request by silasdavis**
It's not very desirable to require clients to make sure they don't write blocking callbacks when subscribing to events (particularly when AddListener doesn't have a comment reflecting this contract). It can lead to some nasty and unexpected bugs downstream and seems unnecessary, for example:

https://github.com/eris-ltd/eris-db/pull/266

Better would be to use a timeout for the callbacks and not hold the eventCell listeners under lock unnecessarily.

This is my first hash at an implementation, may need some work but I think it's an improvement and keeps the same sort of synchronisation guarantees as currently exists (i.e. if an event is fired after I have subscribed I will see it, if I have unsubscribed before an event fires I won't see it).

A simpler approach would be to drop the lock on FireEvent altogether, which might not hurt for some use-cases.

Another approach would be to have separate locks for adding and removing listeners. Then just use the removal mutex to protect the map from concurrent access on delete, but not worry about sending an event to an listener that has just unsubscribed (who cares?).

Still this is probably the neatest if we allow timeouts.

### CODE REVIEW

**Comment by silasdavis** [comment]
Oh I also linked an eventCell to its parent EventSwitch, which isn't strictly necessary but makes the error nicer and I think is good for debugging in general.

I also haven't written unit tests 👿 ... but could be persuaded to when I have time, and if you approve of the general direction of the change.

File: events.go [hunk 1]
```diff
@@ -159,11 +168,28 @@ func (cell *eventCell) RemoveListener(listenerID string) int {
 }
 
 func (cell *eventCell) FireEvent(data EventData) {
+	// Avoid nasty downstream bugs by blocking while taking a snapshot but not
+	// while actually executing callbacks
+	// enforcing non-blocking on event callbacks
 	cell.mtx.RLock()
-	for _, listener := range cell.listeners {
-		listener(data)
+	listenersSnapshot := make(map[string]eventCallback, len(cell.listeners))
```
Comment (silasdavis) [inline events.go]: Not strictly necessary, but let's not hang on to the lock any longer than we need to

**Comment by ebuchman** [comment]
Hmm thanks for this. Sorry for the bug. We've noted the issue for ourselves before at https://github.com/tendermint/tendermint/issues/168. Not sure what the best solution is - will try and review later today.

**Comment by jaekwon** [comment]
We can't create a new goroutine for each FireEvent if for some use-cases, event callbacks need to be called in order.  But we can create a goroutine for each listener.

Right now AddListenerForEvent() takes a callback function.  What if we implement `func RunListener(capacity int, cb eventCallback) eventCallback` that creates a wrapped eventCallback func with its own goroutine?  And add a comment on AddListenerForEvent() that it runs synchronously with FireEvents(), and to use RunListener(x, cb)?

**Comment by silasdavis** [comment]
@jaekwon I'm not sure this buys you much more than telling your consumers 'don't block'. After all it's easy enough for them to launch any blocking operation in a go routine themselves. Are you adverse to the idea of having some timeout as described?

It seems to me the biggest issue with asynchronous execution of callbacks is if the site calling FireEvent expects the callbacks to have run before it moves in, as in the VM. I'm not sure if this is really the case, but I thought it might be. It still feels very wrong that the action of a subscriber can cause a publisher to die, kind of action at a distance. If FireEvent can return before knowing each callback has run then we can just run all the callbacks in a single goroutine, then order is maintained but you can't rely on them having executed.

I think this would be a neater design. If the call site of FireEvent really does need to wait for a callback to execute then it could arrange to receive some completion signal from the specific callbacks in question.

Having said that we could simplify further by expecting any causally related callbacks to just be registered as part of a single callback, and then run each callback in its own goroutine. Presumably order-dependent callbacks must be issuing from some single coordinating authority for them to be aware of each other so we could make it the responsibility of that authority. Failing that we could introduce a 'callback pool', the id of which would be an argument to `AddListenerForEvent`. Callbacks with the same pool id will execute in the order registered and will run in a single goroutine, the caveat being that any failing, blocking, or long running callbacks in the same pool (maybe 'stream' or 'lane' is a better term) will effect ones later in the pool.

I think it would be nice for FireEvent to be non-blocking.

**Comment by zramsay** [comment]
this repo  is deprecated

**Pull request closed without merging**